### PR TITLE
feat: Added Ollama engine via OpenAI api

### PR DESCRIPTION
**Pull request by AtakanTekparmak**
Since [Ollama is compatible with OpenAI API](https://ollama.com/blog/openai-compatibility), I thought we could easily integrate Ollama as a model backend.

## Usage
The usage format is `get_engine("ollama-full_model_string")`, for example.
```python
OLLAMA_MODEL_STRING = "gemma2:9b-instruct-fp16" #Any Ollama model string you have on device
engine = get_engine("ollama-"+OLLAMA_MODEL_STRING)
```

## Changed Files
- [textgrad/engine/__init__.py](https://github.com/zou-group/textgrad/compare/main...AtakanTekparmak:textgrad:main?expand=1#diff-2b70b9ba0757925be695c9fc6c1b3af5284ae5bffd23b0d9e93cb3f969eef9cb)
- [textgrad/engine/openai.py](https://github.com/zou-group/textgrad/compare/main...AtakanTekparmak:textgrad:main?expand=1#diff-1390ce0cdb508201dd7d94e5fe7f3cc1390b42999b5aa908ac0028fa661b10e6)

### CODE REVIEW

**Comment by vinid** [comment]
Thanks so much for this!

I'll review more thoroughly this evening but a first though: would it make sense to create a new class (Ollama) that inherits from the current ChatGPT implementation?

Maybe having the two separated would make future updates a bit easier (or more difficult, if they don't propagate).

**Comment by AtakanTekparmak** [comment]
> Thanks so much for this!
> 
> I'll review more thoroughly this evening but a first though: would it make sense to create a new class (Ollama) that inherits from the current ChatGPT implementation?
> 
> Maybe having the two separated would make future updates a bit easier (or more difficult, if they don't propagate).

I think both options are fine. If we're going for a separate file, separate class etc. then we could also just use the [Ollama Python Library](https://github.com/ollama/ollama-python), where it would have a cleaner API for initiating the client than the OpenAI workaround we have right now.

What we can do, if we like the usage api `get_engine("ollama-model-string")`, then we can merge this for the time being (since it's not a breaking change as far as I've tested) and then open an issue for a full fledged implementation if that is what we want. Then the usage would stay the same and we'd just change the implementation in `textgrad/engine/__init__.py` and add a separate 'textgrad/engine/ollama.py'.

How does that sound?

File: textgrad/engine/openai.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 from .base import EngineLM, CachedEngine
 from .engine_utils import get_image_type_from_bytes
 
+OLLAMA_BASE_URL = 'http://localhost:11434/v1'
```
Comment (vinid) [inline textgrad/engine/openai.py:20]: would it make sense to have this as an env variable? not sure what ppl do with ollama but I guess some might want to change this?

File: textgrad/engine/openai.py [hunk 2]
```diff
@@ -26,6 +27,7 @@ def __init__(
         model_string: str="gpt-3.5-turbo-0613",
         system_prompt: str=DEFAULT_SYSTEM_PROMPT,
         is_multimodal: bool=False,
+        base_url: str=None,
```
Comment (vinid) [inline textgrad/engine/openai.py:30]: can you add a param string explaining that base URL is currently used to support ollama?

**Review by vinid — COMMENTED** [review]
looks good. i left a couple of comments that should be quick to solve.

Would you be so kind to write 1/2 tests to check the objects are instantiated correctly?

I think we can merge this after the edits

**Comment by AtakanTekparmak** [comment]
> looks good. i left a couple of comments that should be quick to solve.
> 
> Would you be so kind to write 1/2 tests to check the objects are instantiated correctly?
> 
> I think we can merge this after the edits

Added instantiation test as well. It should be good to go @vinid. Let me know if there's anything more needed

**Review by vinid — APPROVED** [review]

**Comment by vinid** [comment]
thanks so much for this!!

**Comment by AtakanTekparmak** [comment]
> thanks so much for this!!

Thank you for the fast merge 😁

**Pull request merged**